extension/healthcheckextensionv2/internal/grpc/grpc.go

```diff
@@ -0,0 +1,130 @@
+	component.StatusNone:             healthpb.HealthCheckResponse_NOT_SERVING,
+	component.StatusStarting:         healthpb.HealthCheckResponse_NOT_SERVING,
+	component.StatusOK:               healthpb.HealthCheckResponse_SERVING,
+	component.StatusRecoverableError: healthpb.HealthCheckResponse_SERVING,
+	component.StatusPermanentError:   healthpb.HealthCheckResponse_SERVING,
+	component.StatusFatalError:       healthpb.HealthCheckResponse_NOT_SERVING,
+	component.StatusStopping:         healthpb.HealthCheckResponse_NOT_SERVING,
+	component.StatusStopped:          healthpb.HealthCheckResponse_NOT_SERVING,
+}
+
+func (s *Server) Check(
+	_ context.Context,
+	req *healthpb.HealthCheckRequest,
+) (*healthpb.HealthCheckResponse, error) {
+	st, ok := s.aggregator.AggregateStatus(status.Scope(req.Service), status.Concise)
+	if !ok {
+		return nil, grpcstatus.Error(codes.NotFound, "unknown service")
+	}
+
+	return &healthpb.HealthCheckResponse{
+		Status: s.toServingStatus(st.Event),
+	}, nil
+}
+
+func (s *Server) Watch(req *healthpb.HealthCheckRequest, stream healthpb.Health_WatchServer) error {
+	sub := s.aggregator.Subscribe(status.Scope(req.Service), status.Concise)
+	defer s.aggregator.Unsubscribe(sub)
+
+	var lastServingStatus healthpb.HealthCheckResponse_ServingStatus = -1
+	var failureTimer *time.Timer
+	failureCh := make(chan struct{})
+
+	for {
+		select {
+		case st, ok := <-sub:
+			if !ok {
+				return grpcstatus.Error(codes.Canceled, "Server shutting down.")
+			}
+			var sst healthpb.HealthCheckResponse_ServingStatus
+
+			switch {
+			case st == nil:
+				sst = healthpb.HealthCheckResponse_SERVICE_UNKNOWN
+			case s.componentHealthConfig.IncludeRecoverable &&
+				s.componentHealthConfig.RecoveryDuration > 0 &&
+				st.Status() == component.StatusRecoverableError:
+				if failureTimer == nil {
+					failureTimer = time.AfterFunc(
+						s.componentHealthConfig.RecoveryDuration,
+						func() { failureCh <- struct{}{} },
+					)
+				}
+				sst = lastServingStatus
+				if lastServingStatus == -1 {
+					sst = healthpb.HealthCheckResponse_SERVING
+				}
+			default:
+				if failureTimer != nil {
+					if !failureTimer.Stop() {
+						<-failureTimer.C
+					}
+					failureTimer = nil
+				}
+				sst = s.toServingStatus(st.Event)
+			}
+
+			if lastServingStatus == sst {
+				continue
+			}
+
+			lastServingStatus = sst
+
+			err := stream.Send(&healthpb.HealthCheckResponse{Status: sst})
+			if err != nil {
+				return grpcstatus.Error(codes.Canceled, "Stream has ended.")
+			}
+		case <-failureCh:
+			failureTimer.Stop()
+			failureTimer = nil
+			if lastServingStatus == healthpb.HealthCheckResponse_NOT_SERVING {
+				continue
+			}
+			lastServingStatus = healthpb.HealthCheckResponse_NOT_SERVING
+			err := stream.Send(
+				&healthpb.HealthCheckResponse{
+					Status: healthpb.HealthCheckResponse_NOT_SERVING,
+				},
+			)
+			if err != nil {
+				return grpcstatus.Error(codes.Canceled, "Stream has ended.")
```

Comment:
I believe that `stream.Send` can fail in situations where we would not receive a cancellation via `context.Done`. This follows the same pattern as the grpc health server implementation:
https://github.com/grpc/grpc-go/blob/815e2e2d358ff8d6f5e67098271ccec1a2546961/health/server.go#L99-L102

Reply:
Right, that's why I think the messages need to be different. When someone is reporting an error to you and the message says "Stream has ended", I think it would be useful to differentiate between this case here and the other place where the same message is being used.